### ISSUE TITLE
Reporting Hystrix Latency Percentiles as a Histogram

### ISSUE THREAD

**Issue by jtlisi** [ahus1/prometheus-hystrix#4]
https://github.com/soundcloud/prometheus-hystrix/blob/master/src/main/java/com/soundcloud/prometheus/hystrix/HystrixPrometheusMetricsPublisherCommand.java#L193

Is there any reason why these latency's are published as a gauge rather than a histogram? Is there something about them that doesn't meet the [Promtheus Histogram Spec](https://prometheus.io/docs/concepts/metric_types/#histogram)? If not I would be interested in putting a PR to allow for latency metrics to be reported as a histogram.

**Comment by beorn7** [comment]
I guess this code is older than the histograms in the Prometheus exposition format.

However, if Hystrix is just creating percentiles, the appropriate Prometheus metric type would be a summary (with quantiles).

I'm not very familiar with Hystrix and this repository. @tomcz , the only contributor, does not work at SoundCloud anymore. @tomcz , do you still have intentions to work on this code?

In any case, I don't think there are people at SoundCloud who have the resources to further maintain this. If there is a non-SoundCloud maintainer available (@tomcz or anybody else), it might make most sense to fork this out of the SoundCloud GH org and maintain it there.

**Comment by tomcz** [comment]
@beorn7 I don't really have any intentions on maintaining this code. It was necessary for some of the SoundCloud services that I maintained, but since that is no longer my responsibility I am very happy to hand it over to someone who is interested.

**Comment by tomcz** [comment]
@jtlisi When this code was written Hystrix itself was actually calculating these metrics so we had to treat them as gauges, and pull out the interesting percentiles. In addition, we don't really get the raw values so we cannot use prometheus histograms.

**Comment by jtlisi** [comment]
@tomcz I took a look and see what you are saying. The way it's implemented is fine and there is no way to ideally make it a summary or a histogram. I'll close this.

**Comment by beorn7** [comment]
@tomcz thanks for your answer.

Even with this particular issue closed, it would be great to find a maintainer for this repo. If anybody is interested, please let me know.

**Comment by SimenB** [comment]
This is (unless I\m mistaking something) possible in Hystrix 1.5 and newer by using the streams provided in that version. Would love to see this added 😄

https://github.com/Netflix/Hystrix/wiki/Metrics-and-Monitoring#15x-and-beyond

**Comment by SimenB** [comment]
Kinda like this:

```diff
@@ -10,8 +10,11 @@ import com.netflix.hystrix.HystrixCommandGroupKey;
 import com.netflix.hystrix.HystrixCommandKey;
 import com.netflix.hystrix.HystrixCommandMetrics;
 import com.netflix.hystrix.HystrixCommandProperties;
+import com.netflix.hystrix.metric.HystrixCommandCompletion;
+import com.netflix.hystrix.metric.HystrixCommandCompletionStream;
 import com.netflix.hystrix.strategy.metrics.HystrixMetricsPublisherCommand;
 import com.netflix.hystrix.util.HystrixRollingNumberEvent;
+import io.prometheus.client.Histogram;
 
 import java.util.Map;
 import java.util.TreeMap;
@@ -23,6 +26,13 @@ import java.util.concurrent.Callable;
  */
 public class HystrixPrometheusMetricsPublisherCommand implements HystrixMetricsPublisherCommand {
 
+    private static final Histogram HISTOGRAM = Histogram.build()
+            .name("hystrix_latency_seconds")
+            .help("Hystrix Command latency")
+            .labelNames("command_group", "command_name")
+            .exponentialBuckets(0.001, 1.31, 30)
+            .register();
+
     private final Map<String, String> labels;
     private final boolean exportProperties;
 
@@ -30,6 +40,7 @@ public class HystrixPrometheusMetricsPublisherCommand implements HystrixMetricsP
     private final HystrixCircuitBreaker circuitBreaker;
     private final HystrixCommandProperties properties;
     private final HystrixMetricsCollector collector;
+    private final HystrixCommandKey commandKey;
 
     public HystrixPrometheusMetricsPublisherCommand(
             HystrixMetricsCollector collector, HystrixCommandKey commandKey, HystrixCommandGroupKey commandGroupKey,
@@ -46,6 +57,7 @@ public class HystrixPrometheusMetricsPublisherCommand implements HystrixMetricsP
         this.properties = properties;
         this.collector = collector;
         this.metrics = metrics;
+        this.commandKey = commandKey;
     }
 
     @Override
@@ -166,6 +178,8 @@ public class HystrixPrometheusMetricsPublisherCommand implements HystrixMetricsP
             addGauge("property_value_fallback_isolation_semaphore_max_concurrent_requests", propDesc,
                     () -> properties.fallbackIsolationSemaphoreMaxConcurrentRequests().get());
         }
+
+        setUpHistogram();
     }
 
     private void createCumulativeCountForEvent(String name, final HystrixRollingNumberEvent event) {
@@ -193,4 +207,13 @@ public class HystrixPrometheusMetricsPublisherCommand implements HystrixMetricsP
     private int booleanToNumber(boolean value) {
         return value ? 1 : 0;
     }
+
+    private void setUpHistogram() {
+        final Histogram.Child histogramChild = HISTOGRAM.labels(labels.get("command_group"), labels.get("command_name"));
+
+        HystrixCommandCompletionStream.getInstance(commandKey).observe()
+                .map(HystrixCommandCompletion::getTotalLatency)
+                .map(latency -> latency / 1000f)
+                .forEach(histogramChild::observe);
+    }
 }
```

**Comment by beorn7** [comment]
This might work. However, I don't really know enough about Hystrix to act as a maintainer or code reviewer here (and as it seems, nobody else at SoundCloud does). As said, if somebody wants to take this on, they needed to fork this repo and maintain it outside of the SoundCloud org.

**Comment by ahus1** [comment]
New maintainer on board. I presented a solution based on HystrixCommandCompletionStream at the last KubeCon in Berlin and think well get this to work.

**Comment by tomcz** [comment]
Welcome @ahus1. Thanks for taking over the rains mate.

**Comment by ahus1** [comment]
I've upgraded to Hystrix 1.5. All command counters about the events (successful, fallback, etc.) are now derived from HystrixCommandCompletionStream, as well as the latencies. 

The new values are present using new labels. I've deprecated the old onces (by adding a DEPRECATED to the help text).

There is also a first unit test that shows some names (i.e. hystrix_command_event_success_total and hystrix_command_latency_execute_seconds_count). The test suite also write a full metrics output to the target folder, look for sample.txt.

**Please provide feedback on this implementation. I'd like to get feedback on the naming.** The old labels will stick around for a release or two, but I want to remove the rather sooner than later. 

Once this looks OK, we can pull information from the other streams as well.

One strage thing I found: Hystrix still calculates the rolling stats, but they are a little bit deferred (maybe for performance reasons). Therefore it is difficult to write tests for them.

Example output (from the tests): [sample.txt](https://github.com/ahus1/prometheus-hystrix/files/1174478/sample.txt)

To try it out you can use the 3.2.0-SNAPSHOT release available from Sonartype OSS repo (more on that on the main README.adoc in the repo).

**Comment by ahus1** [comment]
I did some tests and the event type is now a label. With the additional counter hystrix_command_event_total you now create a relative number events successful etc.:

irate(exampleapp_hystrix_command_event_total [1m]) / ignoring(event) group_left irate(exampleapp_hystrix_command_total [1m])

Histograms look like this: 

histogram_quantile(0.9, rate(exampleapp_hystrix_command_latency_total_seconds_bucket [60s]))

**Comment by ahus1** [comment]
Updated sample, now without the old metrics:  [sample.txt](https://github.com/ahus1/prometheus-hystrix/files/1185895/sample.txt)

**Comment by ahus1** [comment]
This has now been released as 3.2.0 on maven central.

**Comment by SimenB** [comment]
Is it possible to customize buckets, and/or which commands we want metrics from?
